src/Diagnostics.zig

```diff
@@ -22,6 +23,10 @@ pub const Message = struct {
             expected: u32,
             actual: u32,
         },
+        incompatible_types: struct {
```

Comment:
This adds 16 bytes to `Message` in release modes, I think it would be better to add a `typePairStr` function to `Parser` and put that in the error message.

`Tag` should also probably be changed to a tagged union.

Reply:
I did the first part in 9d6ef755c6d3066453bef7836527550ee70d8425.

Reply:
I rebased everything onto the latest and got rid of the`incompatible_types` field. It ended up being easier to squash all my commits together since a lot of the changes were intermingled with the latest updates.